src/gpio.c

```diff
@@ -20,8 +20,9 @@ void    gpio_set(GPIO *GPIOx, uint8_t pin, uint8_t value)
 
 void    gpio_select_alternate_function(GPIO *GPIOx, uint8_t pin, uint8_t function)
 {
-    uint8_t offset = (pin % 8) * 4;
+    const uint8_t offset = (pin % 8) * 4;
 
+    GPIOx->MODER = set_bits_with_offset(GPIOx->MODER, pin * 2U, 2, GPIO_MODE_ALTERNATE_FUNC);
```

Comment:
nitpick: It's arguably better to set MODER after AFRL/AFRH, so when you set MODER it instantly switches to the required AF instead of the interim one (whatever happens to be set in AFRL/AFRH).